File: nine-language-version/lessons/06-recursion/recursion.js
#!/usr/bin/env node
/**
 * Lesson 6: Recursion in JavaScript
 *
 * JavaScript supports recursion. Modern engines have some tail call
 * optimization in strict mode, but it's not widely supported.
 * Default stack size varies by engine (~10,000-50,000 calls).
 */

// ====================
// 1. Simple Recursion
// ====================

function factorial(n) {
    // Calculate n! recursively
    if (n <= 1) {  // Base case
        return 1;
    }
    return n * factorial(n - 1);  // Recursive case
}

function factorialTail(n, accumulator = 1) {
    // Tail-recursive factorial
    if (n <= 1) {
        return accumulator;
    }
    return factorialTail(n - 1, n * accumulator);
}

// ====================
// 2. Fibonacci
// ====================

function fibonacci(n) {
    // Calculate nth Fibonacci number (inefficient)
    if (n <= 1) {
        return n;
    }
    return fibonacci(n - 1) + fibonacci(n - 2);
}

function fibonacciMemo(n, memo = {}) {
    // Fibonacci with memoization
    if (n in memo) {
        return memo[n];
    }

    if (n <= 1) {
        return n;
    }

    memo[n] = fibonacciMemo(n - 1, memo) + fibonacciMemo(n - 2, memo);
    return memo[n];
}

// ====================
// 3. Array Recursion
// ====================

function sumArray(arr) {
    // Sum all elements recursively
    if (arr.length === 0) {
        return 0;
    }
    return arr[0] + sumArray(arr.slice(1));
}

function arrayLength(arr) {
    // Calculate length recursively
    if (arr.length === 0) {
        return 0;
    }
    return 1 + arrayLength(arr.slice(1));
}

function reverseArray(arr) {
    // Reverse array recursively
    if (arr.length === 0) {
        return [];
    }
    return [...reverseArray(arr.slice(1)), arr[0]];
}

function maxElement(arr) {
    // Find maximum element recursively
    if (arr.length === 1) {
        return arr[0];
    }

    const restMax = maxElement(arr.slice(1));
    return arr[0] > restMax ? arr[0] : restMax;
}

// ====================
// 4. String Recursion
// ====================

function reverseString(s) {
    // Reverse string recursively
    if (s.length <= 1) {
        return s;
    }
    return reverseString(s.slice(1)) + s[0];
}

function isPalindrome(s) {
    // Check if string is palindrome recursively
    s = s.replace(/\s/g, '').toLowerCase();

    if (s.length <= 1) {
        return true;
    }

    if (s[0] !== s[s.length - 1]) {
        return false;
    }

    return isPalindrome(s.slice(1, -1));
}

// ====================
// 5. Binary Search
// ====================

function binarySearch(arr, target, low = 0, high = arr.length - 1) {
    // Binary search recursively
    if (low > high) {
        return -1;  // Not found
    }

    const mid = Math.floor((low + high) / 2);

    if (arr[mid] === target) {
        return mid;
    } else if (arr[mid] > target) {
        return binarySearch(arr, target, low, mid - 1);
    } else {
        return binarySearch(arr, target, mid + 1, high);
    }
}

// ====================
// 6. Quicksort
// ====================

function quicksort(arr) {
    // Sort array using quicksort
    if (arr.length <= 1) {
        return arr;
    }

    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);

    return [...quicksort(left), ...middle, ...quicksort(right)];
}

// ====================
// 7. Tower of Hanoi
// ====================

function hanoi(n, source, target, auxiliary, moves = []) {
    // Solve Tower of Hanoi puzzle
    if (n === 1) {
        moves.push(`Move disk 1 from ${source} to ${target}`);
        return moves;
    }

    // Move n-1 disks from source to auxiliary
    hanoi(n - 1, source, auxiliary, target, moves);

    // Move largest disk from source to target
    moves.push(`Move disk ${n} from ${source} to ${target}`);

    // Move n-1 disks from auxiliary to target
    hanoi(n - 1, auxiliary, target, source, moves);

    return moves;
}

// ====================
// 8. Tree Traversal
// ====================

class TreeNode {
    constructor(value, left = null, right = null) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
}

function treeHeight(node) {
    // Calculate height of tree recursively
    if (node === null) {
        return 0;
    }

    const leftHeight = treeHeight(node.left);
    const rightHeight = treeHeight(node.right);

    return 1 + Math.max(leftHeight, rightHeight);
}

function treeSum(node) {
    // Sum all values in tree recursively
    if (node === null) {
        return 0;
    }

    return node.value + treeSum(node.left) + treeSum(node.right);
}

function inorderTraversal(node, result = []) {
    // In-order: left, root, right
    if (node === null) {
        return result;
    }

    inorderTraversal(node.left, result);
    result.push(node.value);
    inorderTraversal(node.right, result);

    return result;
}

function preorderTraversal(node, result = []) {
    // Pre-order: root, left, right
    if (node === null) {
        return result;
    }

    result.push(node.value);
    preorderTraversal(node.left, result);
    preorderTraversal(node.right, result);

    return result;
}

function postorderTraversal(node, result = []) {
    // Post-order: left, right, root
    if (node === null) {
        return result;
    }

    postorderTraversal(node.left, result);
    postorderTraversal(node.right, result);
    result.push(node.value);

    return result;
}

// ====================
// 9. Mutual Recursion
// ====================

function isEven(n) {
    // Check if even using mutual recursion
    if (n === 0) {
        return true;
    }
    return isOdd(n - 1);
}

function isOdd(n) {
    // Check if odd using mutual recursion
    if (n === 0) {
        return false;
    }
    return isEven(n - 1);
}

// ====================
// 10. Greatest Common Divisor
// ====================

function gcd(a, b) {
    // GCD using Euclid's algorithm
    if (b === 0) {
        return a;
    }
    return gcd(b, a % b);
}

// ====================
// Tests and Examples
// ====================

function main() {
    console.log("=== Recursion Examples in JavaScript ===\n");

    // Factorial
    console.log("1. Factorial:");
    console.log(`   factorial(5) = ${factorial(5)}`);
    console.log(`   factorialTail(5) = ${factorialTail(5)}`);

    // Fibonacci
    console.log("\n2. Fibonacci:");
    console.log(`   fibonacci(10) = ${fibonacci(10)}`);
    console.log(`   fibonacciMemo(30) = ${fibonacciMemo(30)}`);

    // Array operations
    console.log("\n3. Array Operations:");
    const numbers = [1, 2, 3, 4, 5];
    console.log(`   sumArray(${JSON.stringify(numbers)}) = ${sumArray(numbers)}`);
    console.log(`   arrayLength(${JSON.stringify(numbers)}) = ${arrayLength(numbers)}`);
    console.log(`   reverseArray(${JSON.stringify(numbers)}) = ${JSON.stringify(reverseArray(numbers))}`);
    console.log(`   maxElement(${JSON.stringify(numbers)}) = ${maxElement(numbers)}`);

    // String operations
    console.log("\n4. String Operations:");
    console.log(`   reverseString('hello') = ${reverseString('hello')}`);
    console.log(`   isPalindrome('racecar') = ${isPalindrome('racecar')}`);
    console.log(`   isPalindrome('A man a plan a canal Panama') = ${isPalindrome('A man a plan a canal Panama')}`);

    // Binary search
    console.log("\n5. Binary Search:");
    const sortedArr = [1, 3, 5, 7, 9, 11, 13, 15];
    console.log(`   binarySearch(${JSON.stringify(sortedArr)}, 7) = ${binarySearch(sortedArr, 7)}`);
    console.log(`   binarySearch(${JSON.stringify(sortedArr)}, 4) = ${binarySearch(sortedArr, 4)}`);

    // Quicksort
    console.log("\n6. Quicksort:");
    const unsorted = [3, 6, 8, 10, 1, 2, 1];
    console.log(`   quicksort(${JSON.stringify(unsorted)}) = ${JSON.stringify(quicksort(unsorted))}`);

    // Tower of Hanoi
    console.log("\n7. Tower of Hanoi (3 disks):");
    const moves = hanoi(3, 'A', 'C', 'B');
    moves.forEach(move => console.log(`   ${move}`));

    // Tree operations
    console.log("\n8. Binary Tree:");
    //       5
    //      / \
    //     3   8
    //    / \   \
    //   1   4   9
    const tree = new TreeNode(5,
        new TreeNode(3,
            new TreeNode(1),
            new TreeNode(4)),
        new TreeNode(8,
            null,
            new TreeNode(9)));

    console.log(`   treeHeight() = ${treeHeight(tree)}`);
    console.log(`   treeSum() = ${treeSum(tree)}`);
    console.log(`   inorderTraversal() = ${JSON.stringify(inorderTraversal(tree))}`);
    console.log(`   preorderTraversal() = ${JSON.stringify(preorderTraversal(tree))}`);
    console.log(`   postorderTraversal() = ${JSON.stringify(postorderTraversal(tree))}`);

    // Mutual recursion
    console.log("\n9. Mutual Recursion:");
    console.log(`   isEven(10) = ${isEven(10)}`);
    console.log(`   isOdd(10) = ${isOdd(10)}`);
    console.log(`   isEven(7) = ${isEven(7)}`);
    console.log(`   isOdd(7) = ${isOdd(7)}`);

    // GCD
    console.log("\n10. Greatest Common Divisor:");
    console.log(`   gcd(48, 18) = ${gcd(48, 18)}`);
    console.log(`   gcd(100, 35) = ${gcd(100, 35)}`);
}

if (require.main === module) {
    main();
}

module.exports = {
    factorial,
    factorialTail,
    fibonacci,
    fibonacciMemo,
    sumArray,
    reverseArray,
    binarySearch,
    quicksort,
    TreeNode,
    treeHeight,
    gcd
};

File: nine-language-version/lessons/06-recursion/recursion.rs
#!/usr/bin/env rustc
/**
 * Lesson 6: Recursion in Rust
 *
 * Rust supports recursion but stack space is limited (default ~2-8 MB).
 * Rust does NOT have tail call optimization (yet), so be careful with
 * deep recursion. The type system and ownership model add safety.
 */

// ====================
// 1. Simple Recursion
// ====================

fn factorial(n: u64) -> u64 {
    // Calculate n! recursively
    if n <= 1 {
        1
    } else {
        n * factorial(n - 1)
    }
}

fn factorial_tail(n: u64) -> u64 {
    // Tail-recursive factorial (not optimized in current Rust)
    fn go(n: u64, acc: u64) -> u64 {
        if n <= 1 {
            acc
        } else {
            go(n - 1, n * acc)
        }
    }
    go(n, 1)
}

// ====================
// 2. Fibonacci
// ====================

fn fibonacci(n: u32) -> u64 {
    // Fibonacci (inefficient)
    if n <= 1 {
        n as u64
    } else {
        fibonacci(n - 1) + fibonacci(n - 2)
    }
}

use std::collections::HashMap;

fn fibonacci_memo(n: u32, memo: &mut HashMap<u32, u64>) -> u64 {
    // Fibonacci with memoization
    if let Some(&result) = memo.get(&n) {
        return result;
    }

    let result = if n <= 1 {
        n as u64
    } else {
        fibonacci_memo(n - 1, memo) + fibonacci_memo(n - 2, memo)
    };

    memo.insert(n, result);
    result
}

// ====================
// 3. Slice Recursion
// ====================

fn sum_slice(arr: &[i32]) -> i32 {
    // Sum all elements recursively
    if arr.is_empty() {
        0
    } else {
        arr[0] + sum_slice(&arr[1..])
    }
}

fn slice_length(arr: &[i32]) -> usize {
    // Calculate length recursively
    if arr.is_empty() {
        0
    } else {
        1 + slice_length(&arr[1..])
    }
}

fn reverse_vec(vec: Vec<i32>) -> Vec<i32> {
    // Reverse vector recursively
    if vec.is_empty() {
        return vec;
    }

    let mut result = reverse_vec(vec[1..].to_vec());
    result.push(vec[0]);
    result
}

fn max_element(arr: &[i32]) -> Option<i32> {
    // Find maximum element
    match arr.len() {
        0 => None,
        1 => Some(arr[0]),
        _ => {
            let rest_max = max_element(&arr[1..]).unwrap();
            Some(if arr[0] > rest_max { arr[0] } else { rest_max })
        }
    }
}

// ====================
// 4. String Recursion
// ====================

fn reverse_string(s: &str) -> String {
    // Reverse string recursively
    if s.len() <= 1 {
        return s.to_string();
    }

    let mut chars: Vec<char> = s.chars().collect();
    let first = chars.remove(0);
    let rest: String = chars.into_iter().collect();

    reverse_string(&rest) + &first.to_string()
}

fn is_palindrome(s: &str) -> bool {
    // Check if palindrome
    let clean: String = s.chars()
        .filter(|c| !c.is_whitespace())
        .map(|c| c.to_lowercase().to_string())
        .collect();

    check_palindrome(&clean)
}

fn check_palindrome(s: &str) -> bool {
    if s.len() <= 1 {
        return true;
    }

    let chars: Vec<char> = s.chars().collect();
    if chars[0] != chars[chars.len() - 1] {
        return false;
    }

    let middle: String = chars[1..chars.len()-1].iter().collect();
    check_palindrome(&middle)
}

// ====================
// 5. Binary Search
// ====================

fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    // Binary search recursively
    fn search(arr: &[i32], target: i32, low: usize, high: usize, offset: usize) -> Option<usize> {
        if low > high {
            return None;
        }

        let mid = (low + high) / 2;

        if arr[mid] == target {
            Some(mid + offset)
        } else if arr[mid] > target {
            if mid == 0 {
                None
            } else {
                search(arr, target, low, mid - 1, offset)
            }
        } else {
            search(arr, target, mid + 1, high, offset)
        }
    }

    if arr.is_empty() {
        None
    } else {
        search(arr, target, 0, arr.len() - 1, 0)
    }
}

// ====================
// 6. Quicksort
// ====================

fn quicksort(arr: Vec<i32>) -> Vec<i32> {
    // Sort using quicksort
    if arr.len() <= 1 {
        return arr;
    }

    let pivot = arr[arr.len() / 2];
    let left: Vec<i32> = arr.iter().filter(|&&x| x < pivot).copied().collect();
    let middle: Vec<i32> = arr.iter().filter(|&&x| x == pivot).copied().collect();
    let right: Vec<i32> = arr.iter().filter(|&&x| x > pivot).copied().collect();

    let mut result = quicksort(left);
    result.extend(middle);
    result.extend(quicksort(right));
    result
}

// ====================
// 7. Tower of Hanoi
// ====================

fn hanoi(n: u32, source: &str, target: &str, auxiliary: &str) -> Vec<String> {
    // Solve Tower of Hanoi
    if n == 1 {
        return vec![format!("Move disk 1 from {} to {}", source, target)];
    }

    let mut moves = Vec::new();

    // Move n-1 from source to auxiliary
    moves.extend(hanoi(n - 1, source, auxiliary, target));

    // Move largest from source to target
    moves.push(format!("Move disk {} from {} to {}", n, source, target));

    // Move n-1 from auxiliary to target
    moves.extend(hanoi(n - 1, auxiliary, target, source));

    moves
}

// ====================
// 8. Tree Traversal
// ====================

#[derive(Debug, Clone)]
struct TreeNode {
    value: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(value: i32) -> Self {
        TreeNode {
            value,
            left: None,
            right: None,
        }
    }

    fn with_children(value: i32, left: Option<Box<TreeNode>>, right: Option<Box<TreeNode>>) -> Self {
        TreeNode { value, left, right }
    }
}

fn tree_height(node: &Option<Box<TreeNode>>) -> u32 {
    // Calculate height
    match node {
        None => 0,
        Some(n) => {
            let left_height = tree_height(&n.left);
            let right_height = tree_height(&n.right);
            1 + left_height.max(right_height)
        }
    }
}

fn tree_sum(node: &Option<Box<TreeNode>>) -> i32 {
    // Sum all values
    match node {
        None => 0,
        Some(n) => n.value + tree_sum(&n.left) + tree_sum(&n.right),
    }
}

fn inorder_traversal(node: &Option<Box<TreeNode>>, result: &mut Vec<i32>) {
    // Left, root, right
    if let Some(n) = node {
        inorder_traversal(&n.left, result);
        result.push(n.value);
        inorder_traversal(&n.right, result);
    }
}

fn preorder_traversal(node: &Option<Box<TreeNode>>, result: &mut Vec<i32>) {
    // Root, left, right
    if let Some(n) = node {
        result.push(n.value);
        preorder_traversal(&n.left, result);
        preorder_traversal(&n.right, result);
    }
}

fn postorder_traversal(node: &Option<Box<TreeNode>>, result: &mut Vec<i32>) {
    // Left, right, root
    if let Some(n) = node {
        postorder_traversal(&n.left, result);
        postorder_traversal(&n.right, result);
        result.push(n.value);
    }
}

// ====================
// 9. Mutual Recursion
// ====================

fn is_even(n: u32) -> bool {
    // Check if even using mutual recursion
    if n == 0 {
        true
    } else {
        is_odd(n - 1)
    }
}

fn is_odd(n: u32) -> bool {
    // Check if odd
    if n == 0 {
        false
    } else {
        is_even(n - 1)
    }
}

// ====================
// 10. Greatest Common Divisor
// ====================

fn gcd(a: u32, b: u32) -> u32 {
    // Euclid's algorithm
    if b == 0 {
        a
    } else {
        gcd(b, a % b)
    }
}

// ====================
// Tests and Examples
// ====================

fn main() {
    println!("=== Recursion Examples in Rust ===\n");

    // Factorial
    println!("1. Factorial:");
    println!("   factorial(5) = {}", factorial(5));
    println!("   factorial_tail(5) = {}", factorial_tail(5));

    // Fibonacci
    println!("\n2. Fibonacci:");
    println!("   fibonacci(10) = {}", fibonacci(10));
    let mut memo = HashMap::new();
    println!("   fibonacci_memo(30) = {}", fibonacci_memo(30, &mut memo));

    // Slice operations
    println!("\n3. Slice Operations:");
    let numbers = vec![1, 2, 3, 4, 5];
    println!("   sum_slice({:?}) = {}", numbers, sum_slice(&numbers));
    println!("   slice_length({:?}) = {}", numbers, slice_length(&numbers));
    println!("   reverse_vec({:?}) = {:?}", numbers, reverse_vec(numbers.clone()));
    println!("   max_element({:?}) = {:?}", numbers, max_element(&numbers));

    // String operations
    println!("\n4. String Operations:");
    println!("   reverse_string('hello') = {}", reverse_string("hello"));
    println!("   is_palindrome('racecar') = {}", is_palindrome("racecar"));
    println!("   is_palindrome('A man a plan a canal Panama') = {}",
             is_palindrome("A man a plan a canal Panama"));

    // Binary search
    println!("\n5. Binary Search:");
    let sorted_arr = vec![1, 3, 5, 7, 9, 11, 13, 15];
    println!("   binary_search({:?}, 7) = {:?}", sorted_arr, binary_search(&sorted_arr, 7));
    println!("   binary_search({:?}, 4) = {:?}", sorted_arr, binary_search(&sorted_arr, 4));

    // Quicksort
    println!("\n6. Quicksort:");
    let unsorted = vec![3, 6, 8, 10, 1, 2, 1];
    println!("   quicksort({:?}) = {:?}", unsorted, quicksort(unsorted));

    // Tower of Hanoi
    println!("\n7. Tower of Hanoi (3 disks):");
    let moves = hanoi(3, "A", "C", "B");
    for move_str in moves {
        println!("   {}", move_str);
    }

    // Tree operations
    println!("\n8. Binary Tree:");
    //       5
    //      / \
    //     3   8
    //    / \   \
    //   1   4   9
    let tree = Some(Box::new(TreeNode::with_children(
        5,
        Some(Box::new(TreeNode::with_children(
            3,
            Some(Box::new(TreeNode::new(1))),
            Some(Box::new(TreeNode::new(4))),
        ))),
        Some(Box::new(TreeNode::with_children(
            8,
            None,
            Some(Box::new(TreeNode::new(9))),
        ))),
    )));

    println!("   tree_height() = {}", tree_height(&tree));
    println!("   tree_sum() = {}", tree_sum(&tree));

    let mut inorder = Vec::new();
    inorder_traversal(&tree, &mut inorder);
    println!("   inorder_traversal() = {:?}", inorder);

    let mut preorder = Vec::new();
    preorder_traversal(&tree, &mut preorder);
    println!("   preorder_traversal() = {:?}", preorder);

    let mut postorder = Vec::new();
    postorder_traversal(&tree, &mut postorder);
    println!("   postorder_traversal() = {:?}", postorder);

    // Mutual recursion
    println!("\n9. Mutual Recursion:");
    println!("   is_even(10) = {}", is_even(10));
    println!("   is_odd(10) = {}", is_odd(10));
    println!("   is_even(7) = {}", is_even(7));
    println!("   is_odd(7) = {}", is_odd(7));

    // GCD
    println!("\n10. Greatest Common Divisor:");
    println!("   gcd(48, 18) = {}", gcd(48, 18));
    println!("   gcd(100, 35) = {}", gcd(100, 35));
}
